content/0008-standalone-workflow-execution/rfc.md

```diff
@@ -0,0 +1,294 @@
+want to encourage developers to try that. For this reason we need to expose the
+cluster to the user and ensure they can use standard kubectl commands against
+it.
+
+To ensure there's no confusion as to which cluster someone is accessing to
+perform maintenance, `kubectl` commands against the standalone cluster are
+invoked using `relay cluster kubectl`, which knows how to configure kubectl with
+the proper kubeconfig file.
+
+Obviously someone can choose to populate their global kubeconfig with the
+cluster credentials and change context if they want, but any mistakes caused
+doing so is on them.
+
+### Context switching
+
+Similar to kubectl, we add the ability to switch contexts to the Relay CLI. This
+allows us to setup a `local` context that will allow the relay command to run
+workflows against the local cluster. We add a default context called `relay.sh`
+that points to the production API. Any context that indicates a local
+standalone cluster will not require a login.
+
+#### Bootstrapping
+
+* `relay` creates a data directory in a known location
+* `relay` creates a json configuration file in the data directory
+* k3d creates a 3 node Kubernetes cluster and `relay` ensures it's running and
+  healthy
+* Critical 3rd party resources are deployed to the cluster using `kubectl apply`.
+    - tekton
+    - knative
+* Critical 3rd party resources are deployed to the cluster using Helm
+    - Ambassador
+    - Vault
+* Relay resources are deployed to the cluster using `kubectl apply`
+    - relay-core CRDs
+* Helm chart configuration values are generated and populated into a
+  `values.yaml` and saved to disk inside the `config.json` configuration file
+* Relay services are deployed to the cluster using Helm. See how container
+  image management works below
+    - relay-core
+
+This will provide a cluster capable of running workflows that are issued to the
+cluster as a `Workflow` resource using kubectl. But `relay` knows how to create
+that using workflow YAML by using the same packages for workflow translation the
+API uses.
+
+The last part of the bootstrap phase is running workflows.
+
+* `relay` runs all workflows provided by the user inside a known workflows
+  directory.
+    - We check if a known workflow directory exists
+    - If the directory exists, all YAML files are ran against the cluster
+
+The workflow directory hook is handy so we can further bootstrap a development
+environment for the SaaS or make sure any resources we want will automatically
+exist on the cluster without complicating `relay` any futher. We just let the
+cluster utilize its own ability to run workflows.
+
+#### Image management
+
+The key to `relay` being useful is proper container image management. This is a
+feature that allows a developer to easily write code locally and run it against
+a cluster without committing and waiting for a continuous integration cycle.
+
+Container images for Relay services are initially pulled using the `latest` tag
```

Comment:
We'll have to update our deployment infra to tag this, or find another way to reference these images I think. Also, worth noting that for the SaaS side people will need Docker credentials to pull these images, probably established through `gcloud`.

For the `local` side, have you looked at https://github.com/google/ko at all? I wonder if it could simplify some of the syncing process.

Reply:
So k3d already has a mechanism for syncing and managing images inside the cluster. It basically interacts with the host machine's docker service and uses that to find images first, tarball's them, then stores them in a volume the nodes have access to.

`ko` actually looks like a very good project for this. I think it has the potential to dramatically simplify the management of core images in the local cluster. I'll look into it a bit more. Maybe this is something we can experiment with outside the RFC.

The SaaS infra is a bit more complicated. There's many ways to handle it.

We can script the bulk of the process using the deploy infra, always assuming the user has setup and connected to the GCP account with gcloud. This would utilize the image management commands laid out in the RFC, then it would install and run a workflow that would bootstrap those SaaS components.

We can script individual repos that do the same thing, but lets you pull in each component piecemeal.

I think the important bit here is making sure it's always done by importing local images into the cluster. This allows us to easily build new images from uncommitted code and then run that on the dev cluster by importing the new image and running a workflow that deploys it.